### DISCUSSION TITLE
흥배님 안녕하세요 질문이 있어서 문의 드립니다.

### DISCUSSION THREAD

**Discussion by iluvs** [jacking75/SuperSocketLite#17] [Q&A]
우선 SuperSocketLite를 공유해 주셔서 진심으로 감사하게 생각하고 있습니다.

그런데 저희가 SuperSocketLite로 서버를 구성해서 서비스를 하고 있는데, 가끔씩 AsyncSocketSession.cs 114번째 줄인
ClearPrevSendState 함수에서 ObjectDisposedException이 발생하는데 혹시 이유랑 이 익셉션을 방지할 수 있는 방법이 있을지 문의 드립니다.

**Comment by jacking75** [comment]
안녕하세요 제가 경험해보지 못한 에러인데 에러 발생을 보면 버퍼 데이터의 메모리가 제대로 보존되지 않아서 그런것 같습니다. 
send 함수는 아래와 같을텐데
```
session.Send(reqInfo.Body);
```

send가 끝나기 전에(비동기 이므로 호출과 동시에 끝나지 않습니다) reqInfo.Body의 데이터가 삭제 된 것 같습니다

**Reply by iluvs** [reply]
흥배님 빠르게 답변 주셔서 감사합니다.

네 일단 저희가 작성한 로직쪽을 더 살펴보고 궁금한것이 생기면 또 질문 드리겠습니다.
감사합니다.

**Reply by jacking75** [reply]
네 언제라도 질문 해주시기 바랍니다~

**Reply by iluvs** [reply]
> 네 언제라도 질문 해주시기 바랍니다~

흥배님 안녕하세요.

SuperSocket이나 SuperSocket lite에는 기본적으로 heart beat 기능이 없는지요?
직접 구현해야하는걸까요?

close가 안되고 쌓이고 있어서 직접 만들려고 하는데 혹시 기본적으로 있는 기능이면 사용하려고 합니다.

**Reply by jacking75** [reply]
안녕하세요
heart beat는 애플리케이션 측의 구현이라고 생각해서 기능으로 넣지 않았습니다^^;;;

- 어떤 경우는 keep alive를 사용하는 경우도 있고
-  heart beat 구현 방법이 여러가지가 있어서 어느 것을 원할지 알 수가 없기
때문입니다.

**Reply by iluvs** [reply]
> 안녕하세요
> heart beat는 애플리케이션 측의 구현이라고 생각해서 기능으로 넣지 않았습니다^^;;;
> 
> * 어떤 경우는 keep alive를 사용하는 경우도 있고
> * heart beat 구현 방법이 여러가지가 있어서 어느 것을 원할지 알 수가 없기
>   때문입니다.

그렇군요. 그래서 저도 그냥 heart beat 패킷을 주고 받으면서 지정된 시간안에 응답 안하는 세션은 정리하는 식으로 구현했습니다. 
super socket에서 keep alive를 설정하면 응답 없을 시 그냥 close를 하는 방식인가요? 그렇다면 저도 그걸 그냥 써도 될것 같긴하네요

**Reply by jacking75** [reply]
직접 구현한 것을 사용하시면 됩니다. SuperSocekt은 옵션에서 keep alive 값을 설정할 수 있는데 이것은 커널의 소켓옵션을 사용하는 것입니다. 아직까지 소켓 옵션의 keep alive를 사용하는 것은 추천 받지 못하고 있습니다. heartbeat를 직접 구현하는 것이 좋습니다
